core/src/test/scala/com/evolutiongaming/kafka/flow/TestConsumer.scala

```diff
@@ -0,0 +1,96 @@
+package com.evolutiongaming.kafka.flow
+
+import cats.ApplicativeError
+import cats.data.{NonEmptyMap, NonEmptySet}
+import com.evolutiongaming.skafka.consumer.{
+  Consumer,
+  ConsumerGroupMetadata,
+  OffsetAndTimestamp,
+  RebalanceListener => SRebalanceListener
+}
+import com.evolutiongaming.skafka._
+import scodec.bits.ByteVector
+
+import java.util.regex.Pattern
+import scala.concurrent.duration.FiniteDuration
+
+abstract class TestConsumer[F[_]](implicit F: ApplicativeError[F, Throwable]) extends Consumer[F, String, ByteVector] {
```

Comment:
This is quite large class. What is the reason why we do not want to use simplified `Consumer`? It made the tests more simple and allowed us to very strictly control what methods did we call in a library.

It might be important, because calling some of the methods could cause deadlocks or not work at all.

Reply:
> It might be important, because calling some of the methods could cause deadlocks or not work at all.
All other methods will raise exception, don't see a way of getting dead lock.

The change makes production code more straightforward and simple, for tests we can define separate implementation.

Reply:
Throwing errors in runtime instead of preventing the methods to be called in compile time seems scary.

May be you could give me a small code sample where using `Consumer[F, String, ByteVector]` from `skafka` is more straightforward and simple than using the simplified one, i.e. `Consumer[F]`, so I can also see it?

I did not write that wrapper, it was there in the code before me. I just feel it is a good idea, so there are at least two persons thinking it is a good idea, most probably :blush:

Reply:
> Throwing errors in runtime instead of preventing the methods to be called in compile time seems scary.

You are right that failing in runtime is evil, but hope we can make exception for tests as they always compile & run.

Reply:
Then I do not understand this statement:
> All other methods will raise exception, don't see a way of getting dead lock.

You said it will raise the exceptions while the will not on production.

Reply:
I don't get your last message, sorry.

Reply:
Calling some of the methods in `com.evolutiongaming.skafka.Consumer` can cause a deadlock, because `skafka` tries to call all of them serially. See implementation of `skafka` for more details (https://github.com/evolution-gaming/skafka/blob/master/skafka/src/main/scala/com/evolutiongaming/skafka/consumer/Consumer.scala#L232).

Limiting ourselves to a small subset of `Consumer` functionality is a great way to avoid such a situation and also simplify testing and the code. It is a least-power principle. We only expose what we want to use.

Reply:
> Calling some of the methods in com.evolutiongaming.skafka.Consumer can cause a deadlock, because skafka tries to call all of them serially. See implementation of skafka for more details (https://github.com/evolution-gaming/skafka/blob/master/skafka/src/main/scala/com/evolutiongaming/skafka/consumer/Consumer.scala#L232).

How come implementing **trait** can bring any of implementation details of skafka?

Reply:
> How come implementing trait can bring any of implementation details of skafka?
It is similar to using `Monad` instead of `Sync` despite both being traits. `Sync` has more methods and we want to avoid exposing them to underlying logic.